dev/NavigationView/NavigationView.cpp

```diff
@@ -2132,72 +2395,43 @@ void NavigationView::UpdateNavigationViewUseSystemVisual()
-            list->SetShowFocusVisual(showFocusVisual);
-        }
-        );
+        auto showFocusVisual = ShouldShowFocusVisual();
 
-        PropagateChangeToNavigationViewLists(NavigationViewPropagateTarget::TopListView,
-            [showFocusVisual](NavigationViewList* list)
+        if (IsTopNavigationView())
         {
-            list->SetShowFocusVisual(showFocusVisual);
+            PropagateShowFocusVisualToAllNavigationViewItemsInRepeater(m_leftNavRepeater.get(), showFocusVisual);
         }
-        );
-    }
-}
-
-void NavigationView::SetNavigationViewListPosition(winrt::ListView& listView, NavigationViewListPosition position)
-{
-    if (listView)
-    {
-        if (auto navigationViewList = listView.try_as<winrt::NavigationViewList>())
+        else
         {
-            winrt::get_self<NavigationViewList>(navigationViewList)->SetNavigationViewListPosition(position);
+            PropagateShowFocusVisualToAllNavigationViewItemsInRepeater(m_topNavRepeater.get(), showFocusVisual);
         }
     }
 }
 
-void NavigationView::PropagateNavigationViewAsParent()
-{    
-    PropagateChangeToNavigationViewLists(NavigationViewPropagateTarget::All,
-        [this](NavigationViewList* list)
-            {
-                list->SetNavigationViewParent(*this);
-            }
-        );
-}
-
-void NavigationView::PropagateChangeToNavigationViewLists(NavigationViewPropagateTarget target, std::function<void(NavigationViewList*)> const& function)
+bool NavigationView::ShouldShowFocusVisual()
 {
-    if (NavigationViewPropagateTarget::LeftListView == target || 
-        NavigationViewPropagateTarget::All == target)
-    {
-        PropagateChangeToNavigationViewList(m_leftNavListView.get(), function);
-    }
-    if (NavigationViewPropagateTarget::TopListView == target ||
-        NavigationViewPropagateTarget::All == target)
-    {
-        PropagateChangeToNavigationViewList(m_topNavListView.get(), function);
-    }
-    if (NavigationViewPropagateTarget::OverflowListView == target ||
-        NavigationViewPropagateTarget::All == target)
-    {
-        PropagateChangeToNavigationViewList(m_topNavListOverflowView.get(), function);
-    }
+    return SelectionFollowsFocus() == winrt::NavigationViewSelectionFollowsFocus::Disabled;
 }
 
-void NavigationView::PropagateChangeToNavigationViewList(winrt::ListView const& listView, std::function<void(NavigationViewList*)> const& function)
+void NavigationView::PropagateShowFocusVisualToAllNavigationViewItemsInRepeater(winrt::ItemsRepeater const& ir, bool showFocusVisual)
 {
-    if (listView)
+    if (ir)
     {
-        if (auto navigationViewList = listView.try_as<winrt::NavigationViewList>())
+        if (auto itemsSourceView = ir.ItemsSourceView())
         {
-            auto container = winrt::get_self<NavigationViewList>(navigationViewList);
-            function(container);
+            auto numberOfItems = itemsSourceView.Count();
+            for (int i = 0; i < numberOfItems; i++)
```

Comment:
>int i = 0; i < numberOfItems; i++ [](start = 17, length = 33)

@RanjeshJ  I think there should be a way to iterate through repeaters items using a range based for loop right?

Reply:
Range based for would still be better

---
In reply to: [353488471](https://github.com/microsoft/microsoft-ui-xaml/pull/1683#discussion_r353488471) [](ancestors = 353488471)

Reply:
This is currently not possible because ItemsSourceView does not provide the iterators for range based for to use. It is a good idea to be able to iterated so we could open a separate issue for that.